### ISSUE TITLE
CI hasn't passed in months

### ISSUE THREAD

**Issue by broady** [GoogleCloudPlatform/ruby-docs-samples#218]
A few failures I can see:
https://circleci.com/gh/GoogleCloudPlatform/ruby-docs-samples/1182
```

Google Analytics on Google App Engine
  can track visit event (PENDING: End-to-end tests skipped)

Pending: (Failures listed here are expected and do not affect your suite's status)

  1) Google Analytics on Google App Engine can track visit event
     # End-to-end tests skipped
     # ./spec/google_analytics_spec.rb:29

Finished in 0.0007 seconds (files took 0.44262 seconds to load)
1 example, 0 failures, 1 pending

> gcloud app versions delete -analytics-1182 -q
ERROR: (gcloud.app.versions.delete) unrecognized arguments: -analytics-1182
Usage: gcloud app versions delete VERSIONS [VERSIONS ...] [optional flags]
  optional flags may be  --help | --service

For detailed information on this command and its flags, run:
  gcloud app versions delete --help

Failed to delete e2e version
```

```
Failures:

  1) Cat Friends should display a list of cats
     Failure/Error: raise ActionController::RoutingError, "No route matches [#{env['REQUEST_METHOD']}] #{env['PATH_INFO'].inspect}"
     
     ActionController::RoutingError:
       No route matches [GET] "/assets/application-63d002b4806ba81b3fd0ee343c8d37f5160ede65d821329ab2ce721f09280c3a.css"
     # /opt/circleci/.rvm/gems/ruby-2.3.4/gems/railties-5.1.3/lib/rails/rack/logger.rb:36:in `call_app'
     # /opt/circleci/.rvm/gems/ruby-2.3.4/gems/railties-5.1.3/lib/rails/rack/logger.rb:24:in `block in call'
     # /opt/circleci/.rvm/gems/ruby-2.3.4/gems/railties-5.1.3/lib/rails/rack/logger.rb:24:in `call'
     # /opt/circleci/.rvm/gems/ruby-2.3.4/gems/rack-2.0.3/lib/rack/method_override.rb:22:in `call'
     # /opt/circleci/.rvm/gems/ruby-2.3.4/gems/rack-2.0.3/lib/rack/runtime.rb:22:in `call'
     # /opt/circleci/.rvm/gems/ruby-2.3.4/gems/rack-2.0.3/lib/rack/sendfile.rb:111:in `call'
     # /opt/circleci/.rvm/gems/ruby-2.3.4/gems/railties-5.1.3/lib/rails/engine.rb:522:in `call'
     # /opt/circleci/.rvm/gems/ruby-2.3.4/gems/rack-2.0.3/lib/rack/urlmap.rb:68:in `block in call'
     # /opt/circleci/.rvm/gems/ruby-2.3.4/gems/rack-2.0.3/lib/rack/urlmap.rb:53:in `each'
     # /opt/circleci/.rvm/gems/ruby-2.3.4/gems/rack-2.0.3/lib/rack/urlmap.rb:53:in `call'
     # /opt/circleci/.rvm/gems/ruby-2.3.4/gems/capybara-2.15.1/lib/capybara/server.rb:44:in `call'
     # /opt/circleci/.rvm/gems/ruby-2.3.4/gems/rack-2.0.3/lib/rack/handler/webrick.rb:86:in `service'
     # ------------------
     # --- Caused by: ---
     # Capybara::CapybaraError:
     #   Your application server raised an error - It has been raised in your test code because Capybara.raise_server_errors == true
     #   /opt/circleci/.rvm/gems/ruby-2.3.4/gems/capybara-2.15.1/lib/capybara/session.rb:145:in `raise_server_error!'
```

```

Failures:

  1) Google Cloud Spanner API samples create_database
     Failure/Error:
       create_database project_id:  @project_id,
                       instance_id: @instance.instance_id,
                       database_id: @database_id
     
     Google::Cloud::ResourceExhaustedError:
       8:Unable to create database 'test_database_1512163400' because the instance 'projects/cloud-samples-tests-ruby/instances/cloud-samples-tests-instance' has already reached the maximum database limit (100). Please delete a database in the instance and try again, or choose a different instance.
```

```

Failures:

  1) Detect Faces detect faces from local image file
     Failure/Error:
       expect {
         detect_faces project_id: @project_id,
                      image_path: image_path("face_no_surprise.jpg")
       }.to output(
         /Surprise: false/
       ).to_stdout
     
       expected block to output /Surprise: false/ to stdout, but output "Joy:      true\nAnger:    true\nSorrow:   false\nSurprise: true\n"
       Diff:
       @@ -1,2 +1,5 @@
       -/Surprise: false/
       +Joy:      true
       +Anger:    true
       +Sorrow:   false
       +Surprise: true
       
     # ./spec/detect_faces_spec.rb:34:in `block (2 levels) in <top (required)>'

  2) Detect Faces detect faces from image file in Google Cloud Storage
     Failure/Error:
       expect {
         detect_faces_gcs project_id: @project_id,
                          image_path: storage_file.to_gs_url
       }.to output(
         /Surprise: false/
       ).to_stdout
     
       expected block to output /Surprise: false/ to stdout, but output "Joy:      true\nAnger:    true\nSorrow:   false\nSurprise: true\n"
       Diff:
       @@ -1,2 +1,5 @@
       -/Surprise: false/
       +Joy:      true
       +Anger:    true
       +Sorrow:   false
       +Surprise: true
       
     # ./spec/detect_faces_spec.rb:46:in `block (2 levels) in <top (required)>'

Finished in 30.42 seconds (files took 0.56411 seconds to load)
22 examples, 2 failures
```

```
Failures:

  1) Datastore sample supports inequality_range
     Failure/Error: expect(tasks.size).to eq(1)
     
       expected: 1
            got: 2
     
       (compared using ==)
     # ./spec/sample_spec.rb:345:in `block (2 levels) in <top (required)>'

  2) Datastore sample supports equal_and_inequality_range
     Failure/Error: expect(tasks.size).to eq(1)
     
       expected: 1
            got: 2
     
       (compared using ==)
     # ./spec/sample_spec.rb:360:in `block (2 levels) in <top (required)>'

Finished in 13.13 seconds (files took 0.23901 seconds to load)
60 examples, 2 failures

Failed examples:

rspec ./spec/sample_spec.rb:340 # Datastore sample supports inequality_range
rspec ./spec/sample_spec.rb:355 # Datastore sample supports equal_and_inequality_range
```

**Comment by frankyn** [comment]
I'll review errors after my next meeting. If you have time @bshaffer, I'd appreciate your help!

**Comment by frankyn** [comment]
@broady I started trying to address the issues in PR#211, and haven't completed this task. I'll see if I can green the tests. I'd like to introduce kokoro into the repo, but I haven't prioritized that. I'll see how far I get with fixing tests before I venture down that path. 

If you have tips on how to best deal with this state I'd appreciate your input. Thank you!

**Comment by broady** [comment]
sounds good. i have a lot of things learned from deflaking the Go tests that I should share. a few things:

- retries within the tests
- using a project pool for proper resource segregation
- aggressive cleanup, aided by the project pool
- deterministic resource names, aided by the project pool

but it looks like there's a bunch of low hanging fruit for these ones. for example, it looks like the spanner db's aren't being cleaned up.

**Comment by frankyn** [comment]
Thanks Chris,

- Updated Spanner tests to be deterministic
- Added waiting before asserting an expectation on a logging.
- Updated Gemfile.lock files (need to do this automatically in the future as this is done locally and using a PR). 
- Added additional Ruby versions to test on.
- Fixed a few small errors in a few other samples that made it into the repo.

I'm waiting for tests to finish to get an idea of where CircleCI is now. If you could help with proper resource segregation that would be awesome!

**Comment by frankyn** [comment]
Added retries to the following sample tests:
Storage
Pubsub
Datastore

Next item will be to add kokoro support unless we should introduce project resource pools first? @broady

**Comment by frankyn** [comment]
Status update: Master passed. Pubsub passed after retries kicked in when testing [Ruby 2.4.2](https://circleci.com/gh/GoogleCloudPlatform/ruby-docs-samples/1214?utm_campaign=build-fixed&utm_medium=email&utm_source=notification). Flaky test is `pulls a message`.

**Comment by broady** [comment]
Nice! Let's leave it at that for now.

**Comment by frankyn** [comment]
Sg!

**Comment by frankyn** [comment]
@broady, I recalled that AppEngine e2e tests are not being tested at the moment. The tests currently timeout CircleCI before completing. Reviewing in a follow-up PR.

**Comment by frankyn** [comment]
Ran e2e tests: https://circleci.com/gh/GoogleCloudPlatform/ruby-docs-samples/1225

I think this is a good reason to move over to kokoro as the e2e tests timeout CircleCI. An alternative is to split tests into multiple nodes.

**Comment by broady** [comment]
Do the e2e tests run in parallel?

**Comment by frankyn** [comment]
No, they are not set up that way.

**Comment by broady** [comment]
@tbpg, could you provide an update here, or close if this is resolved? thanks!

**Comment by frankyn** [comment]
@broady progress is being tracked here: https://github.com/GoogleCloudPlatform/ruby-docs-samples/issues/251
